docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1938 @@
+
+-  Package AppV2
+
+   -  package-name: ``app1``
+
+   -  package-version: ``2.0.0``
+
+   -  template-ids: ``pkgId2:mod:T``
+
+If a transaction query is created with a templateId specified as
+``#app1:mod:T``, then the events stream will include events from both
+template-ids: ``pkgId1:mod:T`` and ``pkgId2:mod:T``
+
+Codegen
+-------
+
+For packages that support SCU (i.e. LF1.16), generated code will use
+package-names in place of package-ids in template ids. Retrieved data
+from the ledger will be subject to the upgrade transformations described
+in previous sections.
+
+Concretely, this is implemented as follows:
+
+Java
+~~~~
+
+The classes that are generated for each template and interface contain a
+``TEMPLATE_ID`` field, which, for upgradable packages, will now use a
+package name, rather than a package id. In order to allow determining
+the package id of these packages, we add a new ``PACKAGE_ID`` field to all
+such classes. Upgradable packages will also cause ``PACKAGE_NAME`` and
+``PACKAGE_VERSION`` fields to be present.
+
+TypeScript
+~~~~~~~~~~
+
+The ``templateId`` field on generated template classes has been updated to
+use the package-name as the package qualifier for upgrade compatible
+packages. This will be used for command submission and queries, however
+note that the package qualifier given back in queries will contain the
+package-id, rather than package-name. Generated modules will now also
+give the package “reference”, which will be the package-name for upgrade
+compatible packages, otherwise it’ll be the package-id.
+
+To perform package-id qualified commands/queries in an upgrade
+compatible package, a copy of the template object can be created using
+the following:
+
+.. code:: typescript
+
+  const MyTemplateWithPackageId = { 
+    ...pkg.Mod.MyTemplate,
+    templateId: `${pkg.packageId}:Mod:MyTemplate`,
+  }
+
+.. _json-api-server-1:
+
+JSON API Server
+----------------
+
+As before, template IDs may still be used with a package id, however,
+for packages built as LF 1.16 or greater, the package may also be
+identified by name, i.e. for upgradable packages, a template ID can have
+the form ``#<package-name>:<module-name>:<template-name>``, and this will be
+resolved to corresponding templates from all packages which share this
+name, and are built at 1.16 or above. For packages built at LF 1.15 or
+lower, the templates will not be identifiable via a package name, and a
+package id must be used.
+
+Note: template ids in query results will always use a package id. This
+allows us to distinguish the source of a particular contract. This means
+that if you use a template with a package name in the request, you can
+no longer expect the template ids in the result to exactly match the
+input template id.
+
+Package ID selection preference: preferences apply to JSON API where you
+can specify your preferred selection of package versions.
+
+PQS
+---
+
+TODO: Update this depending on what happens with PQS.
```

Comment:
Note this TODO, which will depend on whether PQS re-allows direct package-id queries. Waiting on answer here - i believe @basvangijzel-DA is checking

Reply:
@raphael-speyer-da as mentioned in Slack, would be good to have an answer here.

Reply:
I'd have to defer to @filmackay on that one.

Reply:
Here is Fil's answer:

We have flipped our Read API to name, not id. This is independent of Ledger API and focused on our use cases where users never want id - even in the <=1.15 target. To preserve id would have left an enormous amount of tech debt in implementing two incompatible systems and disrupt everything including the storage model. If you want id exactly as it was: users are welcome to use the old PQS version as we maintain a comparability matrix

Reply:
Note that this compatibility matrix is not public, for now I have not mentioned it.